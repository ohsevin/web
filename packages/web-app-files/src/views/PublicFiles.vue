<template>
  <div>
    <list-loader v-if="loadResourcesTask.isRunning" />
    <template v-else>
      <not-found-message v-if="folderNotFound" class="files-not-found oc-height-1-1" />
      <no-content-message
        v-else-if="isEmpty"
        id="files-public-list-empty"
        class="files-empty"
        icon="folder"
      >
        <template #message>
          <span v-translate>There are no resources in this folder</span>
        </template>
        <template v-if="currentFolder.canCreate()" #callToAction>
          <span v-translate data-testid="public-files-call-to-action">
            Drag files and folders here or use the "+ New" button to upload
          </span>
        </template>
      </no-content-message>
      <resource-table
        v-else
        id="files-public-files-table"
        v-model="selected"
        class="files-table"
        :class="{ 'files-table-squashed': !sidebarClosed }"
        :are-thumbnails-displayed="displayThumbnails"
        :resources="paginatedResources"
        :target-route="targetRoute"
        :header-position="fileListHeaderY"
        :sort-by="sortBy"
        :sort-dir="sortDir"
        @fileClick="$_fileActions_triggerDefaultAction"
        @rowMounted="rowMounted"
        @sort="handleSort"
      >
        <template #contextMenu="{ resource }">
          <context-actions v-if="isResourceInSelection(resource)" :items="selected" />
        </template>
        <template #footer>
          <pagination :pages="paginationPages" :current-page="paginationPage" />
          <list-info
            v-if="paginatedResources.length > 0"
            class="oc-width-1-1 oc-my-s"
            :files="totalFilesCount.files"
            :folders="totalFilesCount.folders"
            :size="totalFilesSize"
          />
        </template>
      </resource-table>
    </template>
  </div>
</template>

<script>
import { mapGetters, mapActions, mapMutations, mapState } from 'vuex'
import ResourceTable, { determineSortFields } from '../components/FilesList/ResourceTable.vue'
import { useFileListHeaderPosition, usePagination, useSort } from '../composables'
import { useMutationSubscription, useRouteQuery, useStore } from 'web-pkg/src/composables'
import { fileList } from '../helpers/ui'

import MixinAccessibleBreadcrumb from '../mixins/accessibleBreadcrumb'
import MixinFileActions from '../mixins/fileActions'
import MixinMountSideBar from '../mixins/sidebar/mountSideBar'

import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from '../constants'
import debounce from 'lodash-es/debounce'
import omit from 'lodash-es/omit'
import { buildResource } from '../helpers/resources'
import { bus } from 'web-pkg/src/instance'
import { useTask } from 'vue-concurrency'
import { nextTick, computed, unref } from '@vue/composition-api'

import ListLoader from '../components/FilesList/ListLoader.vue'
import NoContentMessage from '../components/FilesList/NoContentMessage.vue'
import NotFoundMessage from '../components/FilesList/NotFoundMessage.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import ContextActions from '../components/FilesList/ContextActions.vue'
import { DavProperties, DavProperty } from 'web-pkg/src/constants'
import { linkRoleUploaderFolder } from '../helpers/share'
import { createLocationOperations, createLocationPublic } from '../router'

// hacky, get rid asap, just a workaround
const unauthenticatedUserReady = async (router, store) => {
  // exit early which could happen if
  // the resources get reloaded
  // another application decided that the user is already provisioned
  if (store.getters.userReady) {
    return
  }

  // pretty low level, error prone and weak, add method to the store to obtain the publicToken
  // it looks like that something was available in the past, store.state.Files.publicLinkInEdit ...
  const publicToken = (router.currentRoute.params.item || '').split('/')[0]
  const publicLinkPassword = store.getters['Files/publicLinkPassword']

  await store.dispatch('loadCapabilities', {
    publicToken,
    ...(publicLinkPassword && { user: 'public', password: publicLinkPassword })
  })

  // ocis at the moment is not able to create archives for public links that are password protected
  // till this is supported by the backend remove it hard as a workaround
  if (publicLinkPassword) {
    store.commit('SET_CAPABILITIES', {
      capabilities: omit(store.getters.capabilities, ['files.archivers']),
      version: store.getters.version
    })
  }

  store.commit('SET_USER_READY', true)
}

const visibilityObserver = new VisibilityObserver()
export default {
  components: {
    ResourceTable,
    ListInfo,
    Pagination,
    ListLoader,
    NoContentMessage,
    NotFoundMessage,
    ContextActions
  },

  mixins: [MixinAccessibleBreadcrumb, MixinFileActions, MixinMountSideBar],

  setup() {
    const store = useStore()
    const { refresh: refreshFileListHeaderPosition, y: fileListHeaderY } =
      useFileListHeaderPosition()

    const storeItems = computed(() => store.getters['Files/activeFiles'] || [])
    const fields = computed(() => {
      return determineSortFields(unref(storeItems)[0])
    })

    const { sortBy, sortDir, items, handleSort } = useSort({
      items: storeItems,
      fields
    })

    const paginationPageQuery = useRouteQuery('page', '1')
    const paginationPage = computed(() => parseInt(String(paginationPageQuery.value)))
    const { items: paginatedResources, total: paginationPages } = usePagination({
      page: paginationPage,
      items,
      sortDir,
      sortBy
    })

    useMutationSubscription(['Files/UPSERT_RESOURCE'], async ({ payload }) => {
      await nextTick()
      fileList.accentuateItem(payload.id)
    })

    const loadResourcesTask = useTask(function* (signal, ref, sameRoute, path = null) {
      ref.CLEAR_CURRENT_FILES_LIST()

      try {
        let resources = yield ref.$client.publicFiles.list(
          path || ref.$route.params.item,
          ref.publicLinkPassword,
          DavProperties.PublicLink
        )

        // Redirect to files drop if the link has role "uploader"
        const sharePermissions = parseInt(
          resources[0].getProperty(DavProperty.PublicLinkPermission)
        )
        if (linkRoleUploaderFolder.bitmask(false) === sharePermissions) {
          ref.$router.replace(
            createLocationPublic('files-public-drop', {
              params: { token: ref.$route.params.item }
            })
          )
          return
        }

        resources = resources.map(buildResource)
        ref.LOAD_FILES({ currentFolder: resources[0], files: resources.slice(1) })

        refreshFileListHeaderPosition()
      } catch (error) {
        ref.SET_CURRENT_FOLDER(null)
        console.error(error)

        if (error.statusCode === 401) {
          ref.redirectToResolvePage()
          return
        }
      }

      // this is a workAround till we have extended the bootProcess
      // if a visitor is able to view the current page
      // the user is ready and the TOO LATE provisioning can start.
      // there is no other way at the moment to find out if:
      // publicLink is password protected
      // public link is viewable
      // so we expect if the user is able to load resources, so he also is ready
      yield unauthenticatedUserReady(ref.$router, ref.$store)
      ref.accessibleBreadcrumb_focusAndAnnounceBreadcrumb(sameRoute)
    }).restartable()

    return {
      fileListHeaderY,
      loadResourcesTask,
      paginatedResources,
      paginationPages,
      paginationPage,
      handleSort,
      sortBy,
      sortDir
    }
  },

  computed: {
    ...mapGetters('Files', [
      'publicLinkPassword',
      'selectedFiles',
      'currentFolder',
      'highlightedFile',
      'currentFolder',
      'totalFilesCount',
      'totalFilesSize'
    ]),
    ...mapGetters(['configuration']),
    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    selected: {
      get() {
        return this.selectedFiles
      },
      set(resources) {
        this.SET_FILE_SELECTION(resources)
      }
    },

    folderNotFound() {
      return this.currentFolder === null
    },

    targetRoute() {
      return { name: this.$route.name }
    },

    displayThumbnails() {
      return !this.configuration.options.disablePreviews
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, sameRoute)
        }
      },
      immediate: true
    }
  },
  beforeDestroy() {
    visibilityObserver.disconnect()
  },

  created() {
    const loadResourcesEventToken = bus.subscribe('app.files.list.load', (path) => {
      this.loadResourcesTask.perform(this, this.$route.params.item === path, path)
    })

    this.$on('beforeDestroy', () => bus.unsubscribe('app.files.list.load', loadResourcesEventToken))
  },

  methods: {
    ...mapActions('Files', ['loadPreview']),
    ...mapMutations('Files', [
      'SET_FILE_SELECTION',
      'SET_CURRENT_FOLDER',
      'LOAD_FILES',
      'CLEAR_CURRENT_FILES_LIST'
    ]),

    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: true,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    },

    redirectToResolvePage() {
      this.$router.push(
        createLocationOperations('files-operations-resolver-public-link', {
          params: { token: this.$route.params.item }
        })
      )
    },

    isResourceInSelection(resource) {
      return this.selected?.includes(resource)
    }
  }
}
</script>
