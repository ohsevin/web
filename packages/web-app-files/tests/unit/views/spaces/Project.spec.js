import GetTextPlugin from 'vue-gettext'
import { RouterLinkStub, shallowMount } from '@vue/test-utils'
import { localVue } from '../views.setup'
import { createStore } from 'vuex-extensions'
import Files from '@/__fixtures__/files'
import mockAxios from 'jest-mock-axios'
import SpaceProject from '../../../../src/views/spaces/Project.vue'
import Vuex from 'vuex'

localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

beforeEach(mockAxios.reset)

afterEach(() => {
  jest.unmock('@files/src/composables/index')
})

window.ResizeObserver =
  window.ResizeObserver ||
  jest.fn().mockImplementation(() => ({
    disconnect: jest.fn(),
    observe: jest.fn(),
    unobserve: jest.fn()
  }))

const selectors = {
  spaceImage: '.space-overview-image',
  markdownContainer: '.markdown-container',
  emptySpace: '#files-space-empty',
  noSpace: '.space-not-found'
}

const spaceMocks = {
  noSpace: {},
  spaceWithReadmeAndImage: {
    id: 1,
    name: 'space',
    special: [
      {
        specialFolder: { name: 'readme' }
      },
      {
        specialFolder: { name: 'image' }
      }
    ]
  },
  spaceWithoutReadmeAndImage: {
    id: 1,
    name: 'space',
    special: []
  }
}

describe('Spaces project view', () => {
  it('should not show anything if space can not be found', async () => {
    mockAxios.request.mockImplementationOnce(() => {
      return Promise.resolve({
        data: spaceMocks.noSpace
      })
    })

    const wrapper = getMountedWrapper()
    await wrapper.vm.loadResourcesTask.last

    expect(wrapper.find(selectors.noSpace).exists()).toBeTruthy()
  })

  describe('space image', () => {
    it('should show if given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper(
        [],
        null,
        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAApgAAAKYB3X3/OAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAANCSURBVEiJtZZPbBtFFMZ/M7ubXdtdb1xSFyeilBapySVU8h8OoFaooFSqiihIVIpQBKci6KEg9Q6H9kovIHoCIVQJJCKE1ENFjnAgcaSGC6rEnxBwA04Tx43t2FnvDAfjkNibxgHxnWb2e/u992bee7tCa00YFsffekFY+nUzFtjW0LrvjRXrCDIAaPLlW0nHL0SsZtVoaF98mLrx3pdhOqLtYPHChahZcYYO7KvPFxvRl5XPp1sN3adWiD1ZAqD6XYK1b/dvE5IWryTt2udLFedwc1+9kLp+vbbpoDh+6TklxBeAi9TL0taeWpdmZzQDry0AcO+jQ12RyohqqoYoo8RDwJrU+qXkjWtfi8Xxt58BdQuwQs9qC/afLwCw8tnQbqYAPsgxE1S6F3EAIXux2oQFKm0ihMsOF71dHYx+f3NND68ghCu1YIoePPQN1pGRABkJ6Bus96CutRZMydTl+TvuiRW1m3n0eDl0vRPcEysqdXn+jsQPsrHMquGeXEaY4Yk4wxWcY5V/9scqOMOVUFthatyTy8QyqwZ+kDURKoMWxNKr2EeqVKcTNOajqKoBgOE28U4tdQl5p5bwCw7BWquaZSzAPlwjlithJtp3pTImSqQRrb2Z8PHGigD4RZuNX6JYj6wj7O4TFLbCO/Mn/m8R+h6rYSUb3ekokRY6f/YukArN979jcW+V/S8g0eT/N3VN3kTqWbQ428m9/8k0P/1aIhF36PccEl6EhOcAUCrXKZXXWS3XKd2vc/TRBG9O5ELC17MmWubD2nKhUKZa26Ba2+D3P+4/MNCFwg59oWVeYhkzgN/JDR8deKBoD7Y+ljEjGZ0sosXVTvbc6RHirr2reNy1OXd6pJsQ+gqjk8VWFYmHrwBzW/n+uMPFiRwHB2I7ih8ciHFxIkd/3Omk5tCDV1t+2nNu5sxxpDFNx+huNhVT3/zMDz8usXC3ddaHBj1GHj/As08fwTS7Kt1HBTmyN29vdwAw+/wbwLVOJ3uAD1wi/dUH7Qei66PfyuRj4Ik9is+hglfbkbfR3cnZm7chlUWLdwmprtCohX4HUtlOcQjLYCu+fzGJH2QRKvP3UNz8bWk1qMxjGTOMThZ3kvgLI5AzFfo379UAAAAASUVORK5CYII='
      )
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.spaceImage).exists()).toBeTruthy()
      expect(wrapper.vm.imageContent).not.toBeUndefined()
      expect(wrapper).toMatchSnapshot()
    })
    it('should not show if not given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithoutReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.spaceImage).exists()).toBeFalsy()
      expect(wrapper.vm.imageContent).toEqual('')
    })

    it('should not show within a resource', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper([], { id: 1 })
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.spaceImage).exists()).toBeFalsy()
    })
  })

  describe('space readme', () => {
    it('should show if given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.markdownContainer).exists()).toBeTruthy()
      expect(wrapper.vm.markdownContent).not.toBeUndefined()
      expect(wrapper).toMatchSnapshot()
    })
    it('should not show if not given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithoutReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.vm.markdownContent).toEqual('')
    })
  })

  describe('resources', () => {
    it('should show empty-message if no resources given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.emptySpace).exists()).toBeTruthy()
    })
    it('should show resources if given', async () => {
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMocks.spaceWithReadmeAndImage
        })
      })

      const wrapper = getMountedWrapper([Files['/'][0]])
      await wrapper.vm.loadResourcesTask.last

      expect(wrapper.find(selectors.emptySpace).exists()).toBeFalsy()
      expect(wrapper.vm.paginatedResources.length).toEqual(1)
    })
  })
})

function getMountedWrapper(spaceResources = [], spaceItem = null, imageContent = '') {
  const $route = { params: { page: 1, item: spaceItem }, meta: { title: 'Space' } }
  const $router = {
    afterEach: jest.fn(),
    currentRoute: {
      name: 'space',
      query: {},
      params: { spaceId: 1 }
    },
    resolve: (r) => {
      return { href: r.name }
    }
  }

  return shallowMount(SpaceProject, {
    localVue,
    stubs: {
      RouterLink: RouterLinkStub
    },
    data: () => {
      return {
        imageContent
      }
    },
    mocks: {
      $route,
      $router,
      $client: {
        files: {
          getFileContents: jest.fn(() => 'fileContent'),
          list: jest.fn(() => spaceResources)
        }
      }
    },

    store: createStore(Vuex.Store, {
      getters: {
        configuration: () => ({
          server: 'https://example.com/',
          options: {
            disablePreviews: true
          }
        })
      },
      modules: {
        Files: {
          namespaced: true,
          mutations: {
            CLEAR_CURRENT_FILES_LIST: jest.fn(),
            CLEAR_FILES_SEARCHED: jest.fn(),
            LOAD_FILES: jest.fn()
          },
          actions: {
            loadIndicators: jest.fn()
          },
          getters: {
            activeFiles: () => spaceResources,
            totalFilesCount: () => ({ files: spaceResources.length, folders: 0 }),
            selectedFiles: () => [],
            totalFilesSize: () => 10,
            pages: () => 1
          }
        }
      }
    })
  })
}
