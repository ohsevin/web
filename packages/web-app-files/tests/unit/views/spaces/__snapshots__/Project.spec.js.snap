// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spaces project view space image should show if given 1`] = `
<div class="space-overview">
  <div>
    <!---->
    <div class="oc-grid oc-px-m oc-mt-m oc-child-width-1-3@s">
      <div>
        <div class="oc-position-relative"><img alt="" src="data:undefined;base64,undefined" class="space-overview-image oc-cursor-pointer"></div>
      </div>
      <div>
        <div class="oc-flex oc-mb-s oc-flex-middle">
          <h1 class="space-overview-name oc-text-truncate">space</h1>
          <oc-button-stub type="button" size="medium" arialabel="Show context menu" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="space-context-btn">
            <oc-icon-stub name="more-2" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
          </oc-button-stub>
          <oc-drop-stub dropid="space-context-drop" toggle="#space-context-btn" position="right-start" mode="click" closeonclick="true" paddingsize="small" options="[object Object]"><input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
            <ul class="oc-list oc-files-context-actions">
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="edit" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Rename
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="pencil" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Change subtitle
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="markdown" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Change description
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="image-add" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Upload new space image
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="forbid-2" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Disable
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="information" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Details
                </oc-button-stub>
              </li>
            </ul>
          </oc-drop-stub>
        </div>
        <!---->
        <div>
          <div class="markdown-container"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <no-content-message-stub icon="folder" id="files-space-empty" class="files-empty"></no-content-message-stub>
</div>
`;

exports[`Spaces project view space readme should show if given 1`] = `
<div class="space-overview">
  <div>
    <!---->
    <div class="oc-grid oc-px-m oc-mt-m oc-child-width-1-3@s">
      <!---->
      <div>
        <div class="oc-flex oc-mb-s oc-flex-middle">
          <h1 class="space-overview-name oc-text-truncate">space</h1>
          <oc-button-stub type="button" size="medium" arialabel="Show context menu" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="space-context-btn">
            <oc-icon-stub name="more-2" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
          </oc-button-stub>
          <oc-drop-stub dropid="space-context-drop" toggle="#space-context-btn" position="right-start" mode="click" closeonclick="true" paddingsize="small" options="[object Object]"><input id="space-image-upload-input" type="file" name="file" multiple="multiple" tabindex="-1" accept="image/*">
            <ul class="oc-list oc-files-context-actions">
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="edit" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Rename
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="pencil" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Change subtitle
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="markdown" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Change description
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="image-add" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Upload new space image
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="forbid-2" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Disable
                </oc-button-stub>
              </li>
              <li class="oc-spaces-context-action oc-py-xs oc-px-s">
                <oc-button-stub type="button" size="medium" variation="passive" appearance="raw" justifycontent="left" gapsize="medium">
                  <oc-icon-stub name="information" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
                  Details
                </oc-button-stub>
              </li>
            </ul>
          </oc-drop-stub>
        </div>
        <!---->
        <div>
          <div class="markdown-container"></div>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <no-content-message-stub icon="folder" id="files-space-empty" class="files-empty"></no-content-message-stub>
</div>
`;
